File: 103.二叉树的锯齿形层序遍历.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;
        queue.offer(root);
        boolean flag = true;
        while(!queue.isEmpty()){
            int size = queue.size();
            ArrayList<Integer> list = new ArrayList<>();
            for(int i = 0; i < size; ++i){
                TreeNode temp = queue.peek();
                queue.poll();
                if(flag){
                    list.add(temp.val);
                } else {
                    list.add(0, temp.val);
                }
                if(temp.left != null)
                    queue.offer(temp.left);
                if(temp.right != null)
                    queue.offer(temp.right);
            }
            flag = !flag;
            res.add(list);
        }
        return res;
    }
}

File: 二叉树/103.二叉树的锯齿形层序遍历.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
 var zigzagLevelOrder = function (root) {
    if (root === null)
        return [];
    let queue = [root];
    let res = [];
    let arr = [];
    let lastNode = root;
    let newNode = null;
    let flag = true;
    while (queue.length) {
        let temp = queue.shift();
        if (flag)
            arr.push(temp.val);
        else
            arr.unshift(temp.val);
        if (temp.left !== null) {
            queue.push(temp.left);
            newNode = temp.left;
        }
        if (temp.right !== null) {
            queue.push(temp.right);
            newNode = temp.right;
        }
        if (lastNode === temp) {
            lastNode = newNode;
            res.push([].concat(arr));
            arr = [];
            flag = !flag;
        }
    }
    return res;
};